### ISSUE TITLE
Changing order of Pages is not reflesting in the footer

### ISSUE THREAD

**Issue by Peterr-K** [PrestaShop/PrestaShop#16134]
When I change order of the pages in Backend -> Design -> Pages, it is not changed on front end - footer via Link Widget module.

Numbers in "Position" column in the backend is changed, but again not reflesting in Link Widget. Maybe by mistake Link Widget is sorting modules by "ID of the page" and not by "Position ID" ?

**Steps to reproduce the issue:**

You need to check that all pages are checked in the link block
![image](https://user-images.githubusercontent.com/16067358/67852258-625dbb00-fb0c-11e9-87dc-d0da74944eb8.png)
1. Go to the BO => Design => Pages => change position of some pages.
2. Check the position of these in the BO => Design => link widget => OK
3. check the FO => Those pages are not well sorted

We can solve this problem by following these steps:
1. Change the position of pages
![image](https://user-images.githubusercontent.com/16067358/67853228-5ffc6080-fb0e-11e9-9744-9f48e37b8052.png)

2. Go to the BO => Design => Link widget page => position => OK
3. check the FO => NOK => Issue

3. Uncheck all pages
![image](https://user-images.githubusercontent.com/16067358/67853281-7f938900-fb0e-11e9-8ce4-16bcb36d7986.png)
4. Save
5. Edit the block again & check all pages
![image](https://user-images.githubusercontent.com/16067358/67853391-c2556100-fb0e-11e9-8a3b-163cf96997f6.png)

6. Save
7. check the FO => OK
![image](https://user-images.githubusercontent.com/16067358/67853423-d305d700-fb0e-11e9-9f84-2a075d4e295b.png)

**Comment by khouloudbelguith** [comment]
Hi @jannyba,

In your case, this issue occurs in the FO?
I tried with PS1.7.6.1. & the Link Widget module v3.0.4.
In the BO => After changing the page's position, in the link widget configuration, pages are well displayed.
I attached a screen record
https://drive.google.com/file/d/1W-bwecKZroteXa-9fsdU7Qvj9nS9oMQp/view
Thanks to check & feedback.

**Comment by Peterr-K** [comment]
@khouloudbelguith Thanks for your time,
what you showed is ok, re-ordered pages are re-ordered in the Link Widget module. But you have disabled all these pages in the module. If you will enable them, in the FO they are not respecting the changed order.
Please check these screenshots and try if you have the same behaviour:
https://www.prestashop.com/forums/topic/1005446-pages-reordering-is-not-reflecting-in-link-widget-module/?tab=comments#comment-3165478

**Comment by khouloudbelguith** [comment]
@jannyba, I manage to reproduce the issue with PS1.7.5.2 & PS1.7.6.1 & the ps_linklist v3.0.4.
Steps to reproduce the issue:

You need to check that all pages are checked in the link block
![image](https://user-images.githubusercontent.com/16067358/67852258-625dbb00-fb0c-11e9-87dc-d0da74944eb8.png)
1. Go to the BO => Design => Pages => change position of some pages.
2. Check the position of these in the BO => Design => link widget => OK
3. check the FO => Those pages are not well sorted

We can solve this problem by following these steps:
1. Change the position of pages
![image](https://user-images.githubusercontent.com/16067358/67853228-5ffc6080-fb0e-11e9-9744-9f48e37b8052.png)

2. Go to the BO => Design => Link widget page => position => OK
3. check the FO => NOK => Issue

3. Uncheck all pages
![image](https://user-images.githubusercontent.com/16067358/67853281-7f938900-fb0e-11e9-8ce4-16bcb36d7986.png)
4. Save
5. Edit the block again & check all pages
![image](https://user-images.githubusercontent.com/16067358/67853391-c2556100-fb0e-11e9-8a3b-163cf96997f6.png)

6. Save
7. check the FO => OK
![image](https://user-images.githubusercontent.com/16067358/67853423-d305d700-fb0e-11e9-9f84-2a075d4e295b.png)

PS the link widget v2.1.6 doesn't work correctly.
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by khouloudbelguith** [comment]
Ping @jolelievre

**Comment by NeOMakinG** [comment]
Order should be handled in backend, can't do something there... What do you think @jolelievre @matks ?

**Comment by Peterr-K** [comment]
Order should be handled in backend, can't do something there... What do you think @jolelievre @matks ?
thank you

**Comment by HartLarsson** [comment]
![image](https://user-images.githubusercontent.com/16102074/80309925-421b6a00-87d8-11ea-8934-2a7ec5b009a6.png)
Order works but links are shown in FO not in order
PS 1.7.6.4

**Comment by matks** [comment]
I explored a bit this issue tonight.

The content of a block inside "Link list module" is stored inside MySQL table `ps_link_block`.

If you add the 5 first Pages to the block to the default configuration, the storage will contain that:
```
{"cms":["1","2","3","4","5"],"product":[false],"static":["contact","sitemap","stores"],"category":[false]}
```

So the problem is that, when you save from the "BO > Design > Link list" page the configuration of the block, it stores the order of these pages like above. Then it displays them following this sort. The sorting of these blocks is being hard written inside the database, any change you perform on "BO > Design > Pages" is ignored.

So what this mean

## What cannot be done

A scenario where you go to "BO > Design > Link list" page and you add 3 Pages to a block, they have position 2, 6 and 9.
Then you go to "BO > Design > Pages" and you modify the Pages positions to 5, 6 and 9.
Then automatically this re-orders the Pages displayed in the footer.

This scenario cannot happen unless a big change in the module data model is done.

## What can be done

A new feature could be added to Link list module, it would allow to sort the items added to a block. It would allow BO user to choose the order to use for this list:

<img width="630" alt="Capture d’écran 2022-03-17 à 21 52 24" src="https://user-images.githubusercontent.com/3830050/158893171-8a08aa2f-efa3-4c46-9ca2-6e6e02c26901.png">

This order would be stored inside MySQL to be displayed exactly as the BO user wanted it.

@MatShir following this exploration I suggest to re-qualify this issue.

AFAIK this is not a bug but a limitation of the current system.

I can suggest to transform this issue into an improvement
- if we want to improve it following my suggestion to allow BO user to sort items manually
- if we want to have the "automatic acknowledgement of Pages positions" feature we need to modify the data model

Or if it does not make sense for the project to follow this path I suggest to close this issue.

But this is not a bug. A bug is a behavior that is not working as expected. From the data model I see, the behavior reported here was never planned or expected.

**Comment by kpodemski** [comment]
@matks 

It's exactly what I was thinking: add a sortable plugin to the list of pages so we can easily change via drag&drop. 

If you have 🟢 from PM I can do it :-)

**Comment by matks** [comment]
Adding the `waiting for PM`

**Comment by boo-code** [comment]
The 2022 diagnosis still holds on the current module. `ps_linklist` 7.0.1 on `dev`: `LinkBlockPresenter::makeCmsLinks()` iterates the stored `content['cms']` array and builds one link per id in that order, and nothing reads `ps_cms.position`. The array order is whatever the configuration form submitted, which is why the uncheck-all / recheck-all ritual works - it discards the stored order and rebuilds it from the choice list.

What makes this a defect rather than a stored preference is that the block's own configuration screen already lists the pages in Design > Pages order, as the 2019 screen recording shows. So the front office contradicts the back office under either design, and that part does not need a PM decision.

Fixed on `dev` by ordering on the pages' own `(id_cms_category, position)`, with the id as a tie-break because `position` is unique only within a category. The loop is otherwise untouched - same `new \CMS((int) $cmsId)`, same `null !== $cms->id && $cms->active` filter - so which links appear cannot change, only their sequence, and `position` is already loaded by the constructor so the sort costs no query.

Measured on a 9.2.0 shop, positions set so Design > Pages reads page ids 5, 1, 2, 3, 4 while the block still stored 1, 2, 3, 4, 5:

| | footer order |
| --- | --- |
| base | 1, 2, 3, 4, 5 |
| branch | 5, 1, 2, 3, 4 |
| base again | 1, 2, 3, 4, 5 |

The third row is the control: restoring the file restores the old order. Set identity checked separately - with one page deactivated both sides render the same four links, compared on the full anchor tuple of page id, block id, `href` and visible text, differing only in sequence. Same with a page moved into a second CMS category: category 1's pages in position order, then the other category's.

On the drag-and-drop proposal from 2022: it stays open and it would supersede this sort, since it gives the block an order of its own instead of following the page order. This is the smaller change and it settles the inconsistency either way; it also makes the migration question easier, because "existing blocks start from the page order" is a better default than "from whatever was submitted". Worth stating the consequence plainly: a shop whose footer order came out of the uncheck/recheck workaround will reorder, because that arrangement is exactly the stored array order this stops honouring. The two orders disagree, so there is no way to keep both.

Separate defect found while measuring, and it affects what a merchant sees here: the widget's template cache is not invalidated by a page change, so the corrected order appears only after the shop cache is cleared. With `PS_SMARTY_CACHE` on, editing `ps_link_block.content` directly left the footer unchanged until `var/cache` was cleared. The module's only invalidation hook is `actionGeneralPageSave`, which is declared in `install-dev/data/xml/hook.xml` but dispatched nowhere in 9.2.x; and CMS positions are written through `PrestaShopAdminController::updateGridPosition()` into `GridPositionUpdater::update()`, raw SQL with no hook and no cache clear, so `actionObjectCmsUpdateAfter` does not fire either. That is a core-side gap plus a module hook with no producer, it applies to any change to a page rather than to ordering, and it needs its own issue.


Credit where it is due, from the duplicates: the same direction was proposed on #29197 in August 2022 by mark-app - add `'position' => $cms->position` to each presented link and sort the list on it - and AureRita reproduced the bug there and asked for a PR that never came. `PrestaShop/ps_linklist#71` is a third thread on the same behaviour, closed as a duplicate of this one in 2020.

Two differences in what is prepared here. The sort happens on the `CMS` objects before they are presented, rather than adding a `position` key to the presented link array, because that array is the template contract and a widget template iterating `links` should not gain a field it does not use. And the order is `(id_cms_category, position, id)` rather than `position` alone: `ps_cms.position` is scoped to its category, so sorting on it by itself interleaves pages from different categories by numbers that are not comparable, and the id is a tie-break because positions are not guaranteed unique within a category either.

**Comment by boo-code** [comment]
Fix opened in the module: PrestaShop/ps_linklist#209. Its `Fixes` is cross-repo, so this issue will not close automatically when it merges.
